Improve audit logging

supertassu opened the pull request:
* Modify `audit_events` model to store metadata, changes made (in JSON),
  client information and a reason (for future use). Make modelId column
  optional to allow passing for listings etc.
* Create file `src/server/util/audit.js` for saving audit log entries. It
  exports an async function named `audit()` that can be used to easily create
  audit log entries. It also has an utility function named `getClientData`
  that can be used to extract client data such as the IP address and user agent
  from an Express request object.
* Change all calls to old `models.AuditEvent.createEvent` to use the new
  `audit()` utility.
* Modify both login handlers to add an event to the audit log when an user logs
  in.
* Create an API endpoint `GET /api/audit-events` that can be used by users with
  the `view audit log` permission. It provides a sortable/filterable listing of
  events in the audit log. By default, the permission is assigned to
  administrators.
* Create tests for said API endpoint and other changes in audit logging
  infrastructure.
* Create a basic frontend for listing the last 250 (default value for API)
  events in the audit log.

Bug: #262

googol commented on scripts/set-password.js:
```diff
@@ -3,6 +3,7 @@ import { models } from '../src/server/models'
 
 function printHelp() {
   console.log('set-password.js <email of the user to set password> <password>')
+  process.exit(1)
```
Since this function now exits the process, maybe it should be renamed `printHelpAndExit` or `exitWithHelpMessage` or something like that to reflect the behavior

supertassu replied:
Done

supertassu replied:
Done

googol commented on src/server/boot/03-audit.js:
```diff
@@ -0,0 +1,41 @@
+        order: [order],
+      })
+
+      await audit({
```
maybe the creation of the audit event should be before the `findAll`, so it would show up in the listing

supertassu replied:
Done

googol commented on src/server/boot/03-offline-login.js:
```diff
@@ -37,12 +38,23 @@ export default function (app) {
+  app.use(
+    '/login/password',
+    passport.authenticate('basic'),
+    async (req, res) => {
```
since this is an async handler function, you should use app.wrap for it, to make sure any exceptions get handled properly by express (it doesn't await promises by itself)

supertassu replied:
Done

googol commented on src/server/models/app-models.js:
```diff
@@ -182,6 +185,48 @@ export default function (db) {
 
   const ParticipantAllergy = db.define('participant_allergy')
 
+  const AuditClientData = db.define(
```
I'd argue that for audit data, it's perfectly ok for us to just inline this data into the main AuditEvent. We want to have minimal processing of the data for speed and to minimize any risk of errors when processing it.

supertassu replied:
Sure, anything works for me. My brain is just used to optimize things like this.

googol commented on src/server/models/app-models.js:
```diff
@@ -182,6 +185,48 @@ export default function (db) {
+        allowNull: false,
+      },
+      ipAddress: {
+        type: Sequelize.BIGINT,
```
Since we don't really need any kind of processing of the IP-address, it's only there for reference, we could just store it as a string and not worry about the ip address library and the ip versions

supertassu replied:
Done. Storing it as binary/number makes filtering based on hostmask much easier but if we need to implement things like that at a scale we can always change it later. I changed it to be a string for now.

googol commented on src/server/models/app-models.js:
```diff
@@ -305,6 +372,26 @@ export default function (db) {
+  })
+
+  AuditEvent.belongsTo(User, {
+    onDelete: 'CASCADE',
```
We definitely do not want the user deletions to cause deletions of audit events, or vice versa. The audit events should live essentially forever, regardless of any data. Usually there wouldn't even be these foreign key relationships on the db level, but we can live with them. We just can't have the events being removed because a user might

supertassu replied:
Does sequelize support relations without creating a foreign key? That would make things much easier.

supertassu replied:
It does and I changed it to that. Now there are no foreign keys and if an user gets deleted for some reason it doesn't affect the audit_events table at all and the frontend just says "unknown user".

googol commented on src/server/server.js:
```diff
@@ -51,6 +52,9 @@ async function boot(app) {
     app.use(expressEnforcesSsl())
+  }
+
+  if (process.env.REDIS_URL) {
```
This kind of a change is a bit dangerous for production environments. Before, you could be sure that in production, if the REDIS_URL wasn't defined, the whole thing would crash and you would notice it. Now however, the app would silently use the in-memory session, which is not suitable for production.

I assume you did this change so that you could use the redis session store at development time? We could add a `|| !appConfig.isDev` here so it still forces you to Redis in prod, but allows you to use it in dev too, or we could just revert these lines

supertassu replied:
Yeah, the point was to allow me to use Redis locally. I changed it to always use redis on prod.

googol commented on src/server/util/audit.js:
```diff
@@ -0,0 +1,57 @@
+}
+
+export async function audit({
+  req, // express request object
```
I think passing in req to this audit function is a bit of a layering issue. It pushes the outer edge (express.js), all the way in with the database.

A better model might be to accept the raw data that is needed out of the request (user id, user agent, ip) and create a helper function that can be used to extract all that out of the request. Then the express route handlers would use that function, and pass the data into audit, or any function in between (like the mass edit function), and only the route handlers would need to care about the structure of the req object

```js
audit({ ...getRequestInfo(req), modelId, ... })
```

supertassu replied:
I had already thought of that, just forgot to actually change this :D Anyways, done now.

supertassu commented:
@googol Thanks for the review, I think I fixed all of your comments. I also updated the commit message to reflect these changes. 

Also I responded to your email.

googol commented on src/server/boot/03-audit.js:
```diff
@@ -0,0 +1,38 @@
+      const filter = JSON.parse(req.query.filter || '{}')
+
+      const where = filter.where || {}
+      const limit = +filter.limit || 250
```
just noticed this now, using `+` to coerce strings into numbers is not something i personally like, it is quite unexpected in most places. We should be improving the way these params are parsed, but for now just using `Number(filter.limit)` would be better here, and similarly on the line below

googol commented on src/server/boot/03-participant.js:
```diff
@@ -105,7 +117,7 @@ export default function (app) {
         req.body.fieldName,
         req.body.newValue,
-        req.user.id,
+        req,
```
this should also be a call to `getClientData`

supertassu replied:
Done

googol commented on src/server/util/audit.js:
```diff
@@ -0,0 +1,41 @@
+  const userId = req.user ? req.user.id : null
+  let ipAddress = req.ip
+
+  // if it includes a ':' and a '.', is's probably a "IPv4-mapped IPv6 addresses"
```
since this is only about gathering data, we might be better off not doing this transformation either. The audit data is not something that is actively looked at, it's something that will get looked at when something's wrong. In that case, all available data will be interesting

googol commented on src/server/models/app-models.js:
```diff
@@ -339,11 +344,11 @@ export default function (db) {
 
   AuditEvent.belongsTo(User, {
-    onDelete: 'SET NULL',
+    constraints: false,
```
:ok_hand:

supertassu replied:
I only now realized that #265 also created foreign keys for `presence_histories.userId` with `onDelete: 'SET NULL',`. Should I make that also not use foreign keys like this one?

googol replied:
In that case we probably do want to keep the foreign key, and I think set null is a good behavior for that case

supertassu commented:
> 
> 
> This looks good now. Seems like you've merged in the cross-env branch even though you have a separate PR open for it as well, was that intentional? We can just merge this and close the other one

I made a separate PR in order to allow reviewing that independently (and to allow using that while working on other branches if this still needed more review). You can either close that or merge it - both will have the same outcome.
